File: Janus/converter.py
import subprocess
from os.path import abspath,join
recording_directory='recordings'

def create_files_map(recording_directory):
    process = subprocess.Popen(['ls', recording_directory],
                     stdout=subprocess.PIPE, 
                     encoding='utf-8',
                     universal_newlines=True,
                     stderr=subprocess.PIPE)
    stdout, stderr = process.communicate()                 
    mjr_files={}
    for file_name in stdout.split('\n'):
        if file_name!='':
            [file_name_without_extension,extension]=file_name.split('.')
            splitted_string=file_name_without_extension.split('-')
            if (extension=='mjr' or extension=='wav') and len(splitted_string)==3:
                [call_id,owner,type]=splitted_string
                if call_id not in mjr_files:
                    mjr_files[call_id]={'call_id':call_id,'files':[{
                        'owner':owner,
                        'type':type,
                        'recording_directory':recording_directory,
                        'extension':extension,
                        'file_name_without_extension':file_name_without_extension,
                        'file_name':file_name}]}
                else:
                    mjr_files[call_id]['files'].append({
                        'owner':owner,
                        'type':type,
                        'recording_directory':recording_directory,
                        'extension':extension,
                        'file_name_without_extension':file_name_without_extension,
                        'file_name':file_name}) 
    return mjr_files

def convert_to_wav_files(files):
    for file in files:
        if file['extension']== 'mjr':
            file_path=abspath(join(file['recording_directory'],file['file_name']))
            target_path=abspath(join(file['recording_directory'],file['file_name_without_extension']))        
            process = subprocess.Popen('janus-pp-rec {file_name} {file_name_without_extension}.wav'.format(file_name=file_path,file_name_without_extension=target_path),
                            stdout=subprocess.PIPE, 
                            encoding='utf-8',
                            universal_newlines=True,
                            shell=True,
                            stderr=subprocess.PIPE)
            stdout, stderr = process.communicate()
            print(stdout)
            print(file_path)
            print(stderr)
            process = subprocess.Popen('rm -rf {file_name}'.format(file_name=file_path),
                            stdout=subprocess.PIPE, 
                            encoding='utf-8',
                            universal_newlines=True,
                            shell=True,
                            stderr=subprocess.PIPE)
            stdout, stderr = process.communicate()
            print(stdout)
            print(stderr)
def merge_wav_files(first_file,second_file,target_path):
    command='ffmpeg -y -i {first_file} -i {second_file} -filter_complex amix=inputs=2:duration=first:dropout_transition=2 {target_path}'
    process = subprocess.Popen(command.format(first_file=first_file,second_file=second_file,target_path=target_path),
                    stdout=subprocess.PIPE, 
                    encoding='utf-8',
                    universal_newlines=True,
                    shell=True,
                    stderr=subprocess.PIPE)
    stdout, stderr = process.communicate()
    print(stdout)
    print(stderr)
    process = subprocess.Popen('rm -f {first_file} {second_file}'.format(first_file=first_file,second_file=second_file),
                    stdout=subprocess.PIPE, 
                    encoding='utf-8',
                    universal_newlines=True,
                    shell=True,
                    stderr=subprocess.PIPE)
    stdout, stderr = process.communicate()
    print(stdout)
    print(stderr)


# main block

files={}
files=create_files_map(recording_directory)
for key,val in files.items():
    convert_to_wav_files(val['files'])
files=create_files_map(recording_directory)
for key,val in files.items():
    [file_one,file_two]=val['files']
    file_one_path=abspath(join(file_one['recording_directory'],file_one['file_name']))
    file_two_path=abspath(join(file_two['recording_directory'],file_two['file_name']))
    target_path=abspath(join(file_two['recording_directory'],val['call_id']))+'.wav'
    
    print(file_one_path)
    print(file_two_path)
    print(target_path)
    merge_wav_files(file_one_path,file_two_path,target_path)    

File: Janus/converter/converter.js
import { spawn } from "child_process";
import _ from "lodash";
import ffmpeg from "fluent-ffmpeg";
import Aigle from "aigle";
import { join } from "path";
import { Subject } from "rxjs";
import { rmSync } from "fs";
const spawnObservable = (cmd, args) => {
  const observable = new Subject();
  const writeObservable = new Subject();
  const runCommand = spawn(cmd, args, { shell: true });
  runCommand.on("close", (code) => {
    observable.complete({ code });
  });
  writeObservable.subscribe({
    next: (data) => {
      runCommand.stdin.write(data);
    },
  });
  runCommand.stdout.on("data", (data) => {
    observable.next(data.toString());
  });
  runCommand.on("error", (err) => {
    observable.error(err);
  });
  return { events: observable.asObservable(), sender: writeObservable };
};

const downMixWavFilesFromMjr = async (wavFilesToProcess, targetDirectoryPath) => {
  return Aigle.eachSeries(wavFilesToProcess, async (wavFile) => {
    if (_.size(wavFile.files) < 2) {
      throw new Error("Files Insufficient for conversion");
    }
    const targetPath = join(targetDirectoryPath, `${wavFile.callerId}.wav`);
    const inputFiles = _.map(wavFile.files, ({ wavFilePath }) => {
      return wavFilePath;
    });
    try {
      await downMixAudioFiles(targetPath, ...inputFiles);
      // _.each(wavFile.files, ({ wavFilePath, filePath }) => {
      //   rmSync(wavFilePath, { recursive: true, force: true });
      //   rmSync(filePath, { force: true, recursive: true });
      // });
    } catch (error) {
      console.error("error occurred in downMixing of wav files", error);
      throw error;
    }
  });
};

const convertMjrFilesToAudioFile = async (targetDirectoryPath, ...mjrFiles) => {
  console.log(mjrFiles);
  const wavFilesToProcess = await Aigle.transform(mjrFiles, async (wavFilesToProcess, filePath) => {
    const fileNameWithoutExtension = _.last(_.split(_.first(_.split(filePath, ".mjr")), "/"));
    const fileNameTokens = _.split(fileNameWithoutExtension, "-");
    const [callerId, owner, type] = fileNameTokens;
    if (_.size(fileNameTokens) !== 3) {
      throw new Error("Invalid mjr file name");
    }
    const wavFilePath = join(targetDirectoryPath, `${fileNameWithoutExtension}.wav`);
    if (!wavFilesToProcess[callerId]) {
      wavFilesToProcess[callerId] = {
        callerId,
        files: [],
      };
    }
    const { events: converter } = spawnObservable(`janus-pp-rec`, [filePath, wavFilePath]);
    try {
      await new Promise((resolve, reject) => {
        converter.subscribe({
          error: (error) => {
            console.log("error occurred in mjr conversion", error);
            reject(error);
          },
          complete: async () => {
            resolve();
          },
        });
      });
      wavFilesToProcess[callerId].files.push({
        filePath,
        ready: true,
        wavFilePath,
        callerId,
        fileNameWithoutExtension,
        owner,
        type,
      });
    } catch (error) {
      wavFilesToProcess[callerId].files.push({
        filePath,
        ready: false,
        wavFilePath,
        callerId,
        fileNameWithoutExtension,
        owner,
        type,
      });
      throw error;
    }
  });
  return downMixWavFilesFromMjr(wavFilesToProcess, targetDirectoryPath);
};

const downMixAudioFiles = (outputFilePath, ...inputFilePaths) => {
  return new Promise((resolve, reject) => {
    const command = ffmpeg();
    _.each(inputFilePaths, (path) => {
      command.addInput(path);
    });
    command
      .complexFilter([`amix=inputs=${_.size(inputFilePaths)}:duration=longest`])
      .output(outputFilePath)
      .on("error", (err) => {
        reject(err);
      })
      .on("end", function (err, stdout, stderr) {
        resolve();
      })
      .run();
  });
};
export { convertMjrFilesToAudioFile };
